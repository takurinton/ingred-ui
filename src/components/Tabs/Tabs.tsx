import * as React from "react";
import * as Styled from "./styled";
import { Tab } from "./internal/Tab";
import useEventCallback from "../../hooks/useEventCallback";

type TabsProps = {
  data: {
    text: string;
    count?: number;
  }[];
  value: any;
  withBadge?: boolean;
  onChange?: (value: any) => void;
};

const Tabs = React.forwardRef<HTMLDivElement, TabsProps>(
  ({ data, value, withBadge = false, onChange }, ref) => {
    const valueToIndex = new Map();
    const tabsRef = React.useRef<HTMLDivElement>(null);
    const childrenRef = React.useRef<HTMLDivElement>(null);
    const [indicatorStyle, setIndicatorStyle] = React.useState({});

    const getTabsMeta = () => {
      const tabsNode = tabsRef.current;
      let tabsMeta;
      if (tabsNode) {
        const rect = tabsNode.getBoundingClientRect();
        tabsMeta = {
          clientWidth: tabsNode.clientWidth,
          scrollLeft: tabsNode.scrollLeft,
          scrollTop: tabsNode.scrollTop,
          scrollWidth: tabsNode.scrollWidth,
          top: rect.top,
          bottom: rect.bottom,
          left: rect.left,
          right: rect.right,
        };
      }

      let tabMeta;
      if (tabsNode && value !== false) {
        if (childrenRef.current) {
          const children = childrenRef.current.children;

          if (children.length > 0) {
            const tab = children[valueToIndex.get(value)];
            tabMeta = tab?.getBoundingClientRect() || null;
          }
        }
      }
      return { tabsMeta, tabMeta };
    };

    const updateIndicatorState = useEventCallback(() => {
      const { tabsMeta, tabMeta } = getTabsMeta();
      let startValue = 0;

      if (tabMeta && tabsMeta) {
        const correction = tabsMeta.scrollLeft;
        startValue = tabMeta.left - tabsMeta.left + correction;
      }

      const newIndicatorStyle = {
        left: startValue,
        width: tabMeta ? tabMeta.width : 0,
      };

      if (isNaN(indicatorStyle["left"]) || isNaN(indicatorStyle["width"])) {
        setIndicatorStyle(newIndicatorStyle);
      } else {
        const start = Math.abs(
          indicatorStyle["left"] - newIndicatorStyle["left"],
        );
        const size = Math.abs(
          indicatorStyle["width"] - newIndicatorStyle["width"],
        );

        if (start >= 1 || size >= 1) {
          setIndicatorStyle(newIndicatorStyle);
        }
      }
    });

    React.useEffect(() => {
      updateIndicatorState();
    });

    React.useEffect(() => {
      const handleResize = () => {
        updateIndicatorState();
      };

      window.addEventListener("resize", handleResize);
      return () => window.removeEventListener("resize", handleResize);
    }, [updateIndicatorState]);

    return (
      <Styled.Container ref={ref}>
        <Styled.ChildContainer ref={tabsRef}>
          <Styled.Indicator
            style={{
              ...indicatorStyle,
            }}
          />
          <Styled.ChildrenContainer ref={childrenRef}>
            {data.map((d, index) => {
              const childValue = d.text ?? index;
              valueToIndex.set(childValue, index);
              return (
                <Tab
                  key={d.text}
                  selected={value === d.text}
                  count={d.count}
                  withBadge={withBadge}
                  text={d.text}
                  onChange={onChange}
                />
              );
            })}
            <Styled.Border />
          </Styled.ChildrenContainer>
        </Styled.ChildContainer>
      </Styled.Container>
    );
  },
);

export default Tabs;
