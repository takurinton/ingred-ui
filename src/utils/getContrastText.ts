import { hexToRgba } from "./hexToRgba";

export function decomposeColor(color: any): any {
  // Idempotent
  if (color.type) {
    return color;
  }

  if (color.charAt(0) === "#") {
    return decomposeColor(hexToRgba(color));
  }

  const marker = color.indexOf("(");
  const type = color.substring(0, marker);

  let values = color.substring(marker + 1, color.length - 1);
  let colorSpace;

  if (type === "color") {
    values = values.split(" ");
    colorSpace = values.shift();
    if (values.length === 4 && values[3].charAt(0) === "/") {
      values[3] = values[3].substr(1);
    }
  } else {
    values = values.split(",");
  }
  values = values.map((value: any) => parseFloat(value));

  return { type, values, colorSpace };
}

export function recomposeColor(color: any) {
  const { type, colorSpace } = color;
  let { values } = color;

  if (type.indexOf("rgb") !== -1) {
    // Only convert the first 3 values to int (i.e. not alpha)
    values = values.map((n: any, i: number) => (i < 3 ? parseInt(n, 10) : n));
  } else if (type.indexOf("hsl") !== -1) {
    values[1] = `${values[1]}%`;
    values[2] = `${values[2]}%`;
  }
  if (type.indexOf("color") !== -1) {
    values = `${colorSpace} ${values.join(" ")}`;
  } else {
    values = `${values.join(", ")}`;
  }

  return `${type}(${values})`;
}

export function hslToRgb(color: any) {
  color = decomposeColor(color);
  const { values } = color;
  const h = values[0];
  const s = values[1] / 100;
  const l = values[2] / 100;
  const a = s * Math.min(l, 1 - l);
  const f = (n: any, k = (n + h / 30) % 12) =>
    l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);

  let type = "rgb";
  const rgb = [
    Math.round(f(0) * 255),
    Math.round(f(8) * 255),
    Math.round(f(4) * 255),
  ];

  if (color.type === "hsla") {
    type += "a";
    rgb.push(values[3]);
  }

  return recomposeColor({ type, values: rgb });
}

function getLuminance(color: any) {
  color = decomposeColor(color);

  let rgb =
    color.type === "hsl"
      ? decomposeColor(hslToRgb(color)).values
      : color.values;
  rgb = rgb.map((val: any) => {
    if (color.type !== "color") {
      val /= 255; // normalized
    }
    return val <= 0.03928 ? val / 12.92 : ((val + 0.055) / 1.055) ** 2.4;
  });

  // Truncate at 3 digits
  return Number(
    (0.2126 * rgb[0] + 0.7152 * rgb[1] + 0.0722 * rgb[2]).toFixed(3),
  );
}

// Use the same logic as
// and mui https://github.com/mui/material-ui/blob/2c70b02e68212c6a708fe05f7a55880e50814607/packages/mui-material/src/styles/createPalette.js#L194-L217
export function getContrastRatio(foreground: any, background: any) {
  const lumA = getLuminance(foreground);
  const lumB = getLuminance(background);
  return (Math.max(lumA, lumB) + 0.05) / (Math.min(lumA, lumB) + 0.05);
}

export function getContrastText(background: any): any {
  const contrastText =
    getContrastRatio(background, "#fff") >= 3 ? "#fff" : "#000";

  if (process.env.NODE_ENV !== "production") {
    const contrast = getContrastRatio(background, contrastText);
    if (contrast < 3) {
      console.log("error");
    }
  }

  return contrastText;
}
