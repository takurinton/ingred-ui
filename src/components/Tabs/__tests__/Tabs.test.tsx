import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import Tabs from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Tabs component testing", () => {
  afterEach(cleanup);

  test("Tabs", () => {
    const options = {
      data: [{ text: "hoge" }, { text: "fuga" }],
    };
    const { asFragment } = renderWithThemeProvider(
      <Tabs {...options} value={"hoge"} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Tabs with badge", () => {
    const options = {
      data: [
        { text: "hoge", count: 1 },
        { text: "fuga", count: 2 },
        { text: "piyo", count: 3 },
      ],
    };
    const { asFragment } = renderWithThemeProvider(
      <Tabs {...options} value={"hoge"} withBadge={true} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
