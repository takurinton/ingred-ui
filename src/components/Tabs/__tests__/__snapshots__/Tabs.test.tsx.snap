// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component testing Tabs 1`] = `
<DocumentFragment>
  <div
    class="sc-guDLey eOSJFX"
  >
    <div
      class="sc-jTQCzO dqquKl"
      style="left: 0px; width: 0px;"
    />
    <div
      class="sc-dmyCSP"
    >
      <button
        class="sc-hLQSwg dzBuAb"
        tabindex="0"
        value="hoge"
      >
        <div
          alignitems="center"
          class="sc-fsYfdN ioOgIH"
          display="flex"
          justifycontent="space-between"
        >
          <p
            align="left"
            class="sc-beySPh kECpqC sc-eDLKkx duKwoa"
            color="#041C33"
            font-size="14px"
            lineheight="1.4"
            weight="normal"
          >
            hoge 
          </p>
        </div>
      </button>
      <button
        class="sc-hLQSwg csgFEf"
        tabindex="-1"
        value="fuga"
      >
        <div
          alignitems="center"
          class="sc-fsYfdN ioOgIH"
          display="flex"
          justifycontent="space-between"
        >
          <p
            align="left"
            class="sc-beySPh kECpqC sc-eDLKkx bmQqny"
            color="#041C33"
            font-size="14px"
            lineheight="1.4"
            weight="normal"
          >
            fuga 
          </p>
        </div>
      </button>
      <div
        class="sc-gLLuof VZdEm"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component testing Tabs with badge 1`] = `
<DocumentFragment>
  <div
    class="sc-guDLey eOSJFX"
  >
    <div
      class="sc-jTQCzO dqquKl"
      style="left: 0px; width: 0px;"
    />
    <div
      class="sc-dmyCSP"
    >
      <button
        class="sc-hLQSwg dzBuAb"
        tabindex="0"
        value="hoge"
      >
        <div
          alignitems="center"
          class="sc-fsYfdN cYcpjR"
          display="flex"
          justifycontent="space-between"
        >
          <p
            align="left"
            class="sc-beySPh kECpqC sc-eDLKkx duKwoa"
            color="#041C33"
            font-size="14px"
            lineheight="1.4"
            weight="normal"
          >
            hoge 
          </p>
          <span
            backgroundcolor="#0B82F4"
            class="sc-iBdnpw jvDida"
            color="#FFFFFF"
            font-size="13px"
            font-weight="bold"
            type="pill"
          >
            1
          </span>
        </div>
      </button>
      <button
        class="sc-hLQSwg csgFEf"
        tabindex="-1"
        value="fuga"
      >
        <div
          alignitems="center"
          class="sc-fsYfdN cYcpjR"
          display="flex"
          justifycontent="space-between"
        >
          <p
            align="left"
            class="sc-beySPh kECpqC sc-eDLKkx bmQqny"
            color="#041C33"
            font-size="14px"
            lineheight="1.4"
            weight="normal"
          >
            fuga 
          </p>
          <span
            backgroundcolor="#778490"
            class="sc-iBdnpw gyNwAl"
            color="#FFFFFF"
            font-size="13px"
            font-weight="bold"
            type="pill"
          >
            2
          </span>
        </div>
      </button>
      <button
        class="sc-hLQSwg csgFEf"
        tabindex="-1"
        value="piyo"
      >
        <div
          alignitems="center"
          class="sc-fsYfdN cYcpjR"
          display="flex"
          justifycontent="space-between"
        >
          <p
            align="left"
            class="sc-beySPh kECpqC sc-eDLKkx bmQqny"
            color="#041C33"
            font-size="14px"
            lineheight="1.4"
            weight="normal"
          >
            piyo 
          </p>
          <span
            backgroundcolor="#778490"
            class="sc-iBdnpw gyNwAl"
            color="#FFFFFF"
            font-size="13px"
            font-weight="bold"
            type="pill"
          >
            3
          </span>
        </div>
      </button>
      <div
        class="sc-gLLuof VZdEm"
      />
    </div>
  </div>
</DocumentFragment>
`;
